src/ProjectTemplates/Microsoft.Extensions.AI.Templates/src/ChatWithCustomData/ChatWithCustomData-CSharp.Web/Services/Ingestion/DataIngestor.cs

```diff
@@ -1,62 +1,37 @@
+    public async Task IngestDataAsync(DirectoryInfo directory, string searchPattern)
     {
-        using var scope = services.CreateScope();
-        var ingestor = scope.ServiceProvider.GetRequiredService<DataIngestor>();
-        await ingestor.IngestDataAsync(source);
-    }
-
-    public async Task IngestDataAsync(IIngestionSource source)
-    {
-        await chunksCollection.EnsureCollectionExistsAsync();
-        await documentsCollection.EnsureCollectionExistsAsync();
-
-        var sourceId = source.SourceId;
-        var documentsForSource = await documentsCollection.GetAsync(doc => doc.SourceId == sourceId, top: int.MaxValue).ToListAsync();
-
-        var deletedDocuments = await source.GetDeletedDocumentsAsync(documentsForSource);
-        foreach (var deletedDocument in deletedDocuments)
+        using var writer = new VectorStoreWriter<string>(vectorStore, dimensionCount: IngestedChunk.VectorDimensions, new()
         {
-            logger.LogInformation("Removing ingested data for {DocumentId}", deletedDocument.DocumentId);
-            await DeleteChunksForDocumentAsync(deletedDocument);
-            await documentsCollection.DeleteAsync(deletedDocument.Key);
-        }
-
-        var modifiedDocuments = await source.GetNewOrModifiedDocumentsAsync(documentsForSource);
-        foreach (var modifiedDocument in modifiedDocuments)
-        {
-            logger.LogInformation("Processing {DocumentId}", modifiedDocument.DocumentId);
-            await DeleteChunksForDocumentAsync(modifiedDocument);
-
-            await documentsCollection.UpsertAsync(modifiedDocument);
-
-            var newRecords = await source.CreateChunksForDocumentAsync(modifiedDocument);
-            await chunksCollection.UpsertAsync(newRecords);
-        }
-
-        logger.LogInformation("Ingestion is up-to-date");
-
-        async Task DeleteChunksForDocumentAsync(IngestedDocument document)
+            CollectionName = IngestedChunk.CollectionName,
+            DistanceFunction = IngestedChunk.VectorDistanceFunction,
+            IncrementalIngestion = true,
+        });
+
+        using var pipeline = new IngestionPipeline<string>(
+            reader: new DocumentReader(directory),
+            chunker: new SemanticSimilarityChunker(embeddingGenerator, new(TiktokenTokenizer.CreateForModel("gpt-4o"))),
+            writer: writer,
+            loggerFactory: loggerFactory);
+
+        await foreach (var result in pipeline.ProcessAsync(directory, searchPattern))
         {
-            var documentId = document.DocumentId;
-            var chunksToDelete = await chunksCollection.GetAsync(record => record.DocumentId == documentId, int.MaxValue).ToListAsync();
-            if (chunksToDelete.Count != 0)
+            if (!result.Succeeded)
             {
-                await chunksCollection.DeleteAsync(chunksToDelete.Select(r => r.Key));
+                logger.LogError("Exception occurred during ingestion: {exception}", result.Exception);
```

Comment:
Since you have already passed `loggerFactory` to the pipeline ctor, the exception will be logged by the pipeline. So here we log it for the second time. To keep it simple we could just use log info and log whether it has succeeded or not for given `result.DocumentId`.